---
title: "GSE32866, Genome-scale DNA methylation profiling of lung adenocarcinoma:"
author: "Paulyna Magana"
output: pdf_document
always_allow_html: true

knit: (function(input_file, encoding) { output_dir = "./PDFs_outcome";
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), output_dir, 'GSE32866.pdf'))})
---

```{r setup, include=FALSE, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, comment = NA)
```

## Introduction

Load the following packages:


```{r load packages, message=FALSE,  warning=FALSE}
library(rmarkdown) # paged table
library(GEOquery) #geo query access
library(dplyr)
library(devtools)
library(ggplot2)
library(limma)
library(Glimma)
library(edgeR)
library(kableExtra)
library(ggrepel)
library(tinytex)
library(purrr)
Sys.setenv(VROOM_CONNECTION_SIZE = 25600000)

```

## Importing the data

```{r importing, message=FALSE, warning=FALSE}

my_id <- "GSE32866"
gse <- getGEO(my_id)
```

Some datasets on GEO may be derived from different microarray platforms. Therefore the object gse is a list of different datasets. You can find out how many were used by checking the length of the gse object. Usually there will only be one platform and the dataset we want to analyse will be the first object in the list (gse[[1]]).

```{r length, message=FALSE, comment=NA}
length(gse)
```

## Extract the data

```{r length2, message=FALSE, comment=NA}
gse <- gse[[1]]
gse
```

## Exploratory analysis

The `exprs` function can retrieve the expression values as a data frame; with one column per-sample and one row per-gene.

```{r, message=FALSE}
pdata= pData(gse) #sample information
edata= exprs(gse) #expression data
fdata = fData(gse) #gene annotation
```

## Make sure dimensions match up

The number of rows of the feature data should match the number of rows of the genomic data (both are the number of genes). The number of rows of the phenotype data should match the number of columns of the genomic data (both are the number of samples).

```{r, comment=NA}
dim(fdata)
```

```{r, comment=NA}
#Dimension of edata
dim(pdata)
```

```{r, comment=NA}
#Dimension of edata
dim(edata)
```

## Look at overall distributions

For visualisation and statistical analysis, we will inspect the data to discover what scale the data are presented in. The methods we will use assume the data are on a log2 scale; typically in the range of 0 to 16.

The `summary` function can then be used to print the distributions.

```{r, message=FALSE, results="asis"}
## exprs get the expression levels as a data frame and get the distribution
dat <- summary(exprs(gse))
m <- matrix(1:ncol(dat), 5)

for (i in 1:ncol(m)) {
 cat(kbl(dat[, m[, i]], 'latex', booktabs=TRUE), "\\newline")
}

```

A boxplot can also be generated to see if the data have been normalised. If so, the distributions of each sample should be highly similar.

```{r, message=FALSE}
boxplot(edata,outline=FALSE)
```



```{r}
kable(table(pdata$characteristics_ch1.1),  booktabs = T)
```


```{r}
kable(table(pdata$characteristics_ch1.6,pdata$characteristics_ch1.1), booktabs = T)
```

```{r}

```


## Inspect the clinical variables

Data submitted to GEO contain sample labels assigned by the experimenters, and some information about the processing protocol. All these data can be extracted by the pData function.
```{r}
sampleInfo <- pData(gse)
```


```{r, message=FALSE}
## source_name_ch1 and characteristics_ch1.1 seem to contain factors we might need for the analysis
sampleInfo <- select(sampleInfo, source_name_ch1,characteristics_ch1.1)
```

```{r}
#rename to more convenient column names
sampleInfo <- rename(sampleInfo, patient=characteristics_ch1.1, group = source_name_ch1)
kable(sampleInfo, longtable = T, booktabs = T, caption = "SampleInfo") %>%
kable_styling(latex_options = c("repeat_header"))
```

## Sample clustering and Principal Components Analysis

Unsupervised analysis is a good way to get an understanding of the sources of variation in the data. It can also identify potential outlier samples.


```{r}
library(pheatmap)
corMatrix <- cor(exprs(gse),use="c")
pheatmap(corMatrix)         
```













