---
title: "GSE75037"
author: "Paulyna Magana"
output:
  pdf_document:
    toc: true
    toc_depth: 2
    number_sections: true
    fig_width: 6
    fig_height: 4
    fig_caption: true
header-includes:
- \usepackage{gb4e}
- \noautomath

always_allow_html: true

knit: (function(input_file, encoding) { output_dir = "./PDFs_outcome";
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), output_dir, 'GSE75037.pdf'))})
---

```{r setup, include=FALSE, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, comment = NA)
```

# Introduction

Load the following packages:


```{r load packages, message=FALSE,  warning=FALSE}
library(rmarkdown) # paged table
library(GEOquery) #geo query access
library(knitr) #tables
library(dplyr)
library(devtools)
library(ggplot2)
library(limma)
library(edgeR)
library(ggrepel)
library(kableExtra)
library(tinytex)
Sys.setenv(VROOM_CONNECTION_SIZE = 25600000)
```


# Importing the data

GSE75037, 83 lung adenocarcinomas and 83 matched adjacent non-malignant lung were profiled on Illumina WG6-V3 expression arrays

```{r importing dataset, warning=FALSE, comment=NA, message=FALSE}
my_id <- "GSE75037"
gse <- getGEO(my_id)[[1]]
```

```{r inspect gse, warning=FALSE, comment=NA}
gse
```

# Exploratory analysis

Extract the different tables into "pdata"(pheno data), "edata" (expression data) and "fdata" (features)

```{r extract data}
pdata= pData(gse) #sample information, pheno
edata= exprs(gse) #expression data
fdata = fData(gse) #gene annotation, features
```

# Dimensions

The number of rows of the feature data should match the number of rows of the genomic data (both are the number of genes). The number of rows of the phenotype data should match the number of columns of the genomic data (both are the number of samples).

## Features data

```{r dimensions fdata, comment=NA}
dim(fdata)
```
48803 genes, 28 observations from that gene

## Phenotype data

```{r dimensions pdata, comment=NA}
#Dimension of edata
dim(pdata)
```

## Expression data

166 samples, 50 observations from patient/sample

```{r dimension edata, comment=NA}
#Dimension of edata
dim(edata)
```

48803 genes and 166 samples

# Inspect data

```{r samples}
# Use option useNA to include NA's in table
kable(table(pdata$characteristics_ch1,useNA="ifany"), booktabs=TRUE) %>%
  kable_styling(latex_options = "hold_position")
```

Exactly 83 samples from Adenocarcinoma samples and Non-malignant samples.

```{r female male patients, comment=NA}
kable(table(pdata[43]), booktabs=T)%>%
  kable_styling(latex_options= "hold_position")
```

From the 166 samples, 118 are female patients and 48 male patients.

```{r}
kable(table(pdata$characteristics_ch1, pdata$characteristics_ch1.2), booktabs = T)%>%
  kable_styling(latex_options = "hold_position")
```

# Overall distributions

Data is inspected to discover the scale presented in the expression table.

A boxplot can also be generated to see if the data have been normalised. If so, the distributions of each sample should be highly similar.

```{r boxplot, fig.cap="Boxplot for sample distribution", message=FALSE, warning=FALSE}
boxplot(edata,outline=FALSE)
```



# Inspect the clinical variables

Data submitted to GEO contain sample labels assigned by the experimenters, and some information about the processing protocol. All these data can be extracted by the pData function.

```{r}
sampleInfo <- pdata
## source_name_ch1 and characteristics_ch1.1 seem to contain factors we might need for the analysis
sampleInfo <- select(sampleInfo, source_name_ch1,characteristics_ch1.2)
#rename to more convenient column names
sampleInfo <- rename(sampleInfo, patient=characteristics_ch1.2, group = source_name_ch1)
```

# Sample clustering and Principal Components Analysis

Unsupervised analysis is a good way to get an understanding of the sources of variation in the data. It can also identify potential outlier samples.


```{r}
library(pheatmap)
corMatrix <- cor(exprs(gse),use="c")
pheatmap(corMatrix)         
```

We can incorporate sample information onto the plot to try and understand the clustering. We have already created such a data frame previously (pdata). However, we need to take care that the rownames of these data match the columns of the correlation matrix.

```{r, comment=NA}
rownames(sampleInfo) <- colnames(corMatrix)
pheatmap(corMatrix,
         annotation_col=sampleInfo)
```



```{r MDS, comment=NA, fig.height=6, fig.width=6}
plotMDS(gse, labels=sampleInfo[,"group"],
        gene.selection="common")
```

```{r, warning=FALSE}
#Plot principal components labeled by group
pca <- prcomp(t(edata))
## Join the PCs to the sample information
cbind(sampleInfo, pca$x) %>% 
ggplot(aes(x = PC1, y=PC2, col=group,label=paste("Patient", patient))) + geom_point() + geom_text_repel()
```


# Differential Expression

By far the most-popular package for performing differential expression is limma.
we have to allocate the samples in our dataset to the sample groups of interest. A useful function is  `model.matrix`, which will create a *design matrix* from one of the columns in your `sampleInfo`.

The design matrix is a matrix of `0` and `1`s; one row for each sample and one column for each sample group. A `1` in a particular row and column indicates that a given sample (the row) belongs to a given group (column).

```{r, message=FALSE}
design <- model.matrix(~0+sampleInfo$group)
colnames(design) <- c("adenocarcinoma","non_malignant") #rename columns
kable(head(design), booktabs=T)%>%
  kable_styling(latex_options = "hold_position")
```

In order to perform the differential analysis, we have to define the
#contrast that we are interested in. In our case we only have two groups
#and one contrast of interest.

```{r cutoff, warning=FALSE, message=FALSE}
#calculate median expression level
cutoff <- median(exprs(gse)) 
#TRUE or FALSE whether each gene is "expressed" in each sample
is_expressed <- exprs(gse) >cutoff
## Identify genes expressed in more than 2 samples
keep <- rowSums(is_expressed) >2
## subset to just those expressed genes
gse <- gse[keep,]
```

## Fit the model

```{r fit model, message=FALSE, warning=FALSE}
fit <- lmFit(exprs(gse), design)
```

## Define and fit contrasts

To perform differential analysis, define contrast interested in.

```{r, warning=FALSE}
contrasts <- makeContrasts(adenocarcinoma - non_malignant, levels=design)
fit2 <- contrasts.fit(fit,contrasts)
efit <- eBayes(fit2) #apply empirical Bayes to get the differential expression statistics and p-values
kable(topTable(efit), booktabs=T)%>%
  kable_styling(latex_options = "hold_position")
```

If we want to know how many genes are deferentially-expressed overall

```{r, warning=FALSE}
results <- (decideTests(efit))
kable(summary(results), booktabs=T)
```

Significance is defined using an adjusted p-value cutoff that is set at 5% by default comparison between expression levels in adenocarcinoma and non_mallignant

8622 genes are found to be down-regulated in adenocarcinoma relative to non_malignant
10214 are up-regulated in adenocarcinoma relative to non_malignant

### Add genes IDs

```{r, message=FALSE}
anno <- fData(gse)
anno <- select(anno,Symbol, Entrez_Gene_ID,Chromosome,Cytoband)
efit$genes <- anno
kable(topTable(efit), booktabs=T)%>%
  kable_styling(latex_options= c("scale_down", "hold_position"))
```

## Volcano plot

```{r volcano plot, message=FALSE}
full_results <- topTable(efit, number=Inf)
full_results <- tibble::rownames_to_column(full_results, "ID")
ggplot(full_results, aes(x=logFC, y=B)) +
  geom_point()
```

```{r volcano_plot_pink, warning=FALSE}
p_cutoff <- 0.05
fc_cutoff <- 1
topN <- 20

full_results %>%
  mutate(Significant = adj.P.Val < p_cutoff, abs(logFC) > fc_cutoff) %>%
  mutate(Rank= 1:n(), Label= ifelse(Rank < topN, Symbol, "")) %>%
  ggplot(aes(x = logFC, y=B, col=Significant, label= Label)) +
  geom_text_repel(col="black")
```


```{r plotSA, message=FALSE}
#visualise
plotSA(efit, main="Final model: Mean-variance trend")
```

# Summarize results

Quick look at differential expression levels
```{r, message=FALSE}
kable(topTable(efit), booktabs=T)%>%
  kable_styling(latex_options=c("scale_down", "hold_position"))
```


Some studies require more than an adjusted p-value cut-off for a stricter definition on significance, one may require log-fold-changes (log-FCs) to be above a minimum value.

```{r, message=FALSE}
tfit <- treat(efit, lfc=1)
dt <- decideTests(tfit)
kable(summary(dt), booktabs=T)%>%
  kable_styling(latex_options= "hold_position")
```


```{r, message=FALSE}
adeno_vs_non <- topTreat(tfit, coef=1, n=Inf)
kable(head(adeno_vs_non), booktabs=T) %>%
  kable_styling(latex_options= c("scale_down", "hold_position"))
```


```{r, message=FALSE}
plotMD(tfit, column=1, status=dt[,1], main=colnames(tfit)[1], 
       xlim=c(2,16))
```

## Filtering results

```{r filter_22A1, warning=FALSE}
kable(filter(full_results, Symbol == "SLC22A1"), booktabs=T)%>%
  kable_styling(latex_options=c("scale_down", "hold_position"))
```

```{r filter_22A4, warning=FALSE}
kable(filter(full_results, Symbol == "SLC22A4"), booktabs=T)%>%
  kable_styling(latex_options=c("scale_down", "hold_position"))
```

```{r filter_22A5, warning=FALSE}
kable(filter(full_results, Symbol == "SLC22A5"), booktabs=T)%>%
  kable_styling(latex_options=c("scale_down", "hold_position"))
```

```{r filter_all_22A, warning=FALSE}
kable(filter(full_results, grepl("SLC22", Symbol)), booktabs=T)%>%
  kable_styling(latex_options=c("scale_down", "hold_position"))
```

## Export results

```{r export results, warning=FALSE, message=FALSE}
library(readr)
filter(full_results, adj.P.Val < 0.05, abs(logFC) >1) %>%
  write.csv(file ="./results/filtered_GSE75037.csv")
```

## Session Info

```{r, comment=NA}
sessionInfo()
```

